processing/src/main/java/io/druid/segment/DimensionSelector.java

```diff
@@ -42,7 +44,12 @@
    *
    * Value cardinality would be 2.
    *
-   * @return the value cardinality
+   * Cardinality may be unknown (e.g. the selector used by IncrementalIndex while reading input rows),
+   * in which case this method will return -1. If cardinality is unknown, you should assume this
+   * dimension selector has no dictionary, and avoid calling "lookupId" or "lookupName" outside of
```

Comment:
@drcrallen I updated all the call sites. The groupBy ones didn't seem to really need the cardinality (they were just using it as a marker for missing rows) so I switched them to use an invalid value (`MISSING_VALUE`) instead.

The topN ones did need it, so I adjusted them to check the return value. To make that simpler I reduced the number of call sites by removing "cardinality" as a constructor param to TopNParams. It was always set to the dimension selector's getValueCardinality anyway.